src/libstore/build/drv-output-substitution-goal.cc

```diff
@@ -61,20 +61,21 @@ void DrvOutputSubstitutionGoal::tryNext()
 
     // FIXME: Make async
     // outputInfo = sub->queryRealisation(id);
-    outPipe.create();
-    promise = decltype(promise)();
+
+    downloadState = std::make_shared<DownloadState>();
+    downloadState->outPipe.create();
 
     sub->queryRealisation(
-        id, { [&](std::future<std::shared_ptr<const Realisation>> res) {
+        id, { [downloadState(downloadState)](std::future<std::shared_ptr<const Realisation>> res) {
```

Comment:
I would put some of the info in the code comment as it useful to describing the current state of the code, not just its relationship to the prior states of the code.
```suggestion
        id,
        // The curl download can outlive the `DrvOutputSubstitutionGoal` (if some
        // other error occurs). The by-value capture gives the closure has its
        // own `std::shared_ptr`, which ensures the pointed-to information will live
        // long enough.
        { [downloadState(downloadState)](std::future<std::shared_ptr<const Realisation>> res) {
```

Reply:
Thanks, I 've moved/tweaked the comment.